services/payments/statemachine.go

```diff
@@ -213,6 +214,28 @@ func (s *Service) DriveTransaction(
 			state.LastError = nil
 		}
 
+		// Get the validated history of the transaction from qldb so that we can safely access
+		// current persisted state
+		history, err := s.datastore.GetPaymentStateHistory(ctx, state.DocumentID)
+		if err != nil {
+			return fmt.Errorf("failed to get history from document id", err)
+		}
+		persistedState, err := history.GetAuthenticatedPaymentState(
+			s.verifierStore,
+			state.DocumentID,
+		)
+		if err != nil {
+			return fmt.Errorf("failed to validate payment state history", err)
+		}
+		// If there is idempotency data in qldb and it is different from the idempotency data in the
+		// current state it means that there was a race between two calls to Authenticate and we are
+		// operating on the loser. There is no risk to proceeding as long as we retain the winner
+		// idempotency.
+		if !bytes.Equal(state.ExternalIdempotency, persistedState.ExternalIdempotency) &&
+			persistedState != nil {
+			state.ExternalIdempotency = persistedState.ExternalIdempotency
+		}
```

Comment:
Seems like it's safer to check `persistedState` to not be `nil`, and then access the `ExternalIdempotency` field. 😉 

```suggestion
		if persistedState != nil && !bytes.Equal(state.ExternalIdempotency, persistedState.ExternalIdempotency) {
			state.ExternalIdempotency = persistedState.ExternalIdempotency
		}
```